src/tito/builder/submodule_aware_builder.py

```diff
@@ -0,0 +1,210 @@
+    Usage:
+    # cwd = 'current'
+    with PushDir("into"):
+        # cwd = 'current/into'
+        do_command # cwd is into.
+    # cwd = 'current'
+    """
+
+    def __init__(self, subdir):
+        self.old_wd = os.getcwd()
+        self.new_wd = None
+        if subdir:
+            self.new_wd = os.path.join(self.old_wd, subdir)
+
+    def __enter__(self):
+        if self.new_wd:
+            self.chdir(self.old_wd, self.new_wd)
+
+    def __exit__(self, type, value, traceback):
+        if self.new_wd:
+            self.chdir(self.new_wd, self.old_wd)
+
+    def chdir(self, old_dir, new_dir):
+        if new_dir:
+            debug('Changing directory from %s to %s' % (old_dir, new_dir))
+            os.chdir(new_dir)
+
+
+class SubmoduleAwareBuilder(Builder):
+    """
+    Submodule aware builder class.
+
+    This builder is aware of submodules and will package the contents of the submodules as well.
+    It expectes `git submodule--helper list` to exist and work.
+    """
+    REQUIRED_ARGS = []
+
+    # TODO: drop version
+    def __init__(self, name=None, tag=None, build_dir=None,
+                 config=None, user_config=None,
+                 args=None, **kwargs):
+        # call the parent super class and let it configure everything
+        Builder.__init__(self, name=name, build_dir=build_dir, config=config,
+                         user_config=user_config, args=args, **kwargs)
+        """
+        name - Package name that is being built.
+
+        version - Version and release being built.
+
+        tag - The git tag being built.
+
+        build_dir - Temporary build directory where we can safely work.
+
+        config - Merged configuration. (global plus package specific)
+
+        user_config - User configuration from ~/.titorc.
+
+        args - Optional arguments specific to each builder. Can be passed
+        in explicitly by user on the CLI, or via a release target config
+        entry. Only for things which vary on invocations of the builder,
+        avoid using these if possible.  *Given in the format of a dictionary
+        of lists.*
+        """
+
+    def _setup_sources(self):
+        """
+        Create a copy of the git source for the project at the point in time
+        our build tag was created.
+
+        Created in the temporary rpmbuild SOURCES directory.
+
+        *This is the only thing that SubmoduleAwareBuilder changes*.
+        *It needs to call a different create_tgz
+        """
+        self._create_build_dirs()
+
+        debug("Creating %s from git tag: %s..." % (self.tgz_filename,
+                                                   self.git_commit_id))
+
+        # call our create_tgz, instead of the global one.
+        self.create_tgz(self.git_root, self.tgz_dir, self.git_commit_id,
+                        self.relative_project_dir,
+                        os.path.join(self.rpmbuild_sourcedir, self.tgz_filename))
```

Comment:
What if we update the `Builder._setup_sources` method and change 

```
create_tgz(self.git_root, self.tgz_dir, self.git_commit_id,
           self.relative_project_dir,
           os.path.join(self.rpmbuild_sourcedir, self.tgz_filename))
```

to

```
self._create_tgz()
```

and then define it in the `Builder` class like this

```
def _create_tgz(self):
        return create_tgz(self.git_root, self.tgz_dir, self.git_commit_id,
                          self.relative_project_dir,
                          os.path.join(self.rpmbuild_sourcedir, self.tgz_filename))
```

You could then simply override this method instead of the whole `_setup_sources`. Or if you prefer any other approach, feel free, this is just a suggestion.

Reply:
That would be a lot better, I didn't want to mess with the code that was already there.

If you would like to do the refactor it will be faster.
For me, I'll have to get approval for any changes, and that can take some time.  :)